marklogic-client-api/src/main/java/com/marklogic/client/document/DocumentWriteOperation.java

```diff
@@ -85,4 +90,88 @@
    * @return the logical temporal document URI
    */
   String getTemporalDocumentURI();
+  
+  /**
+   * The from method prepares each content object for writing as a document including generating a URI by inserting a UUID.
+   * @param a subclass of AbstractWriteHandle
+   * @param DocumentUriMaker which internally accepts an AbstractWriteHandle and returns a String
+   * @return a stream of DocumentWriteOperation to be written in the database.
+   */
+    public static Stream<DocumentWriteOperation> from(Stream<? extends AbstractWriteHandle> content,
+            final DocumentUriMaker uriMaker) {
+        final class DocumentWriteOperationImpl implements DocumentWriteOperation {
+            public DocumentWriteOperationImpl(AbstractWriteHandle content, String uri) {
+                this.content = content;
+                this.uri = uri;
+            }
+
+            private AbstractWriteHandle content;
+            private String uri;
+
+            @Override
+            public OperationType getOperationType() {
+                return null;
+            }
+
+            @Override
+            public String getUri() {
+                return uri;
+            }
+
+            @Override
+            public DocumentMetadataWriteHandle getMetadata() {
+                return null;
+            }
+
+            @Override
+            public AbstractWriteHandle getContent() {
+                return content;
+            }
+
+            @Override
+            public String getTemporalDocumentURI() {
+                return null;
+            }
+
+        }
+        final class WrapperImpl {
+            DocumentWriteOperation mapper(AbstractWriteHandle content) {
+                String uri = uriMaker.apply(content);
+                if (uri == null)
+                    throw new MarkLogicInternalException("Uri could not be created");
+                return new DocumentWriteOperationImpl(content, uri);
+            }
+
+        }
+        WrapperImpl wrapperImpl = new WrapperImpl();
```

Comment:
I'm wondering whether a potential improvement on our whiteboarding session would be to pass uriMaker to a WrapperImpl constructor that initialized a private field of WrapperImpl instead of making the uriMaker parameter on the enclosing method final.

That way, WrapperImpl would contain everything needed to make URIs, which might be easier to understand and maintain.  

What do you think?